clients/filesystem-hadoop/src/main/java/com/datastrato/gravitino/filesystem/hadoop/GravitinoVirtualFileSystem.java

```diff
@@ -0,0 +1,389 @@
+            + "/"
+            + identifier.name();
+    if (meta.getFileset().storageLocation().endsWith("/")) {
+      filesetPrefix += "/";
+    }
+    return filesetPrefix;
+  }
+
+  private Path resolvePathByIdentifier(NameIdentifier identifier, FilesetMeta meta, Path path) {
+    String originPath = path.toString();
+    if (originPath.startsWith(GravitinoVirtualFileSystemConfiguration.GVFS_FILESET_PREFIX)) {
+      try {
+        Path storageLocation = new Path(meta.getFileset().storageLocation());
+        boolean isMountSingleFile = meta.getFileSystem().isFile(storageLocation);
+        if (isMountSingleFile) {
+          Preconditions.checkArgument(
+              originPath.equals(concatFilesetPrefix(identifier, meta)),
+              "Cannot resolve path: %s to actual storage path, because the fileset only mounts a single file.",
+              path);
+          return storageLocation;
+        } else {
+          return new Path(
+              originPath.replaceFirst(
+                  concatFilesetPrefix(identifier, meta), storageLocation.toString()));
+        }
+      } catch (Exception e) {
+        throw new RuntimeException(
+            String.format("Cannot resolve path: %s to actual storage path, exception:", path), e);
+      }
+    } else {
+      throw new InvalidPathException(
+          String.format(
+              "Path %s doesn't start with the scheme \"%s\"",
+              path, GravitinoVirtualFileSystemConfiguration.GVFS_FILESET_PREFIX));
+    }
+  }
+
+  private FileStatus resolveFileStatusPathPrefix(
+      FileStatus fileStatus, String fromPrefix, String toPrefix) {
+    String uri = fileStatus.getPath().toString();
+    if (!uri.startsWith(fromPrefix)) {
+      throw new InvalidPathException(
+          String.format("Path %s doesn't start with prefix \"%s\"", uri, fromPrefix));
+    }
+    String proxyUri = uri.replaceFirst(fromPrefix, toPrefix);
+    Path path = new Path(proxyUri);
+    fileStatus.setPath(path);
+    return fileStatus;
+  }
+
+  private NameIdentifier extractIdentifier(URI uri) {
+    Preconditions.checkArgument(
+        uri.toString().startsWith(GravitinoVirtualFileSystemConfiguration.GVFS_FILESET_PREFIX),
+        "Path %s doesn't start with scheme \"%s\"");
+
+    if (StringUtils.isBlank(uri.toString())) {
+      throw new InvalidPathException("Uri which need be extracted cannot be null or empty");
+    }
+
+    // remove first '/' symbol with empty string
+    String[] reservedDirs =
+        Arrays.stream(uri.getPath().replaceFirst("/", "").split("/")).toArray(String[]::new);
+    Preconditions.checkArgument(
+        reservedDirs.length >= 3, "URI %s doesn't contains valid identifier", uri);
+
+    return NameIdentifier.ofFileset(
+        metalake.name(), reservedDirs[0], reservedDirs[1], reservedDirs[2]);
+  }
+
+  private FilesetMeta getCachedFileset(NameIdentifier identifier) throws IOException {
+    try {
+      rwLock.readLock().lock();
```

Comment:
I would suggest moving this line out of `try/catch` block.  `rwLock.readLock().unlock();` will be executed wrongly if `rwLock.readLock().lock()` fails.

Reply:
Done

Reply:
You may also need to change the the locks in method `close`.